### DISCUSSION TITLE
Suggestion/Discussion: How to overcome buffer glitches with some clients

### DISCUSSION THREAD

**Discussion by lars18th** [catalinii/minisatip#1333] [General]
Hi @catalinii ,

As I'm using one Panasonic TV as the SAT>IP client, I'm very worried about the current behaviour of the minisatip with the buffers lengths. I know that with clients with large buffers, the current implementation of minisatip is valid. But I want to discuss with you one idea to overcome the troubles with current buffers and the clients that use small receive buffers.

- After a lot of time using minisatip+Panasonic, I see that even using very small buffers (i.e. `-b 13160:5775360`) some artifacts appear in some cases. And I feel I reduced the cases to this one: when the bitrate decreases rapidly. You can observe this when a frame consumes a very low space in the original stream. So my conclusion is that this small frame, or a chunk of consecutive small frames, are pending in the buffer until they will be processed. In fact, the problem can be described as this: at some time point, a chunk of the initial TS packets of one frame are sended to the client, but the rest of ending packets of the same frame continue in the buffer for a long time.

I noticed that you use a flush of packets with a 50ms period to overcome these troubles. But I feel this stragety is not sufficient, and it can be improved. So my idea is this:

- Imagine that inside the buffer a new video packet with the PUSI mark (typically a start of a new frame, but not true in all cases, specifically with codecs like H.264) is _inserted_ inside the DVR buffer. Then you can assume that all _previous_ packets until this new incoming packet are part of a previous frame. So, the best stratagy is to flush these packets even if the buffer is not completed to the point to process it.

The idea with this is to try to flush "complete" PES packets in a short period of time, and overcome the case when the end of a frame (or a chunk of small frames) are pending in the buffer.

What you think?
Please, comment your opinion.
Regards.

**Comment by catalinii** [comment]
Hey @lars18th,

The current design of the part that reads the data from the DVR buffer is:
1) signal when there is more data available
2) read from the dvr socket
3) call read_dmx which buffers until certain rules are met
4) process_dmx is called: flush all the data to all the clients (makes sure at the end of this the adapter buffer is empty)

On fast systems, I would assume 1 and 2 are called for each packet. The issue is that you do not know in read_dmx when the last packet of the video data before PUSI arrives so you can send that stream. Of course you can always buffer 1316 bytes to make sure the whole UDP packet is filled and send that. If this does not work I would assume there is a network issue (UDP is not correcting errors). 

From personal experience, 50 ms is just a threshold over which the data is flushed. Eg you want to scan a transponder so the data amount is not large, which means you need to flush every 50ms to make sure the scanning progresses. 
For H264 HD channels with suck small buffers (13160 bytes) at 1.3MB/s you will do at least 100 process_dmx'es per second, which means the threshold is 10ms. Considering we are watching 60p videos, the flush happens more often than a frame.

**Comment by lars18th** [comment]
Hi @catalinii ,

Thank you for the response. I'll do more tests based on your comments.

Regards.
